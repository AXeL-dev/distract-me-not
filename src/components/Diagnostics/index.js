export { default as Diagnostics } from './index.jsx';
export { default } from './index.jsx';
