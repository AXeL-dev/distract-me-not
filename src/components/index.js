export * from './Background';
export * from './Blocked';
export * from './Allowed';
export * from './Panel';
export * from './PasswordPrompt';
export * from './AddWebsitePrompt';
export * from './PasteBin';
export * from './Settings';
export * from './Timer';
export * from './Logs';
export * from './Main';
export * from './shared';
