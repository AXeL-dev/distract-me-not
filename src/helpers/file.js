export function download(blob, filename) {
  if (window.navigator.msSaveOrOpenBlob) {
    // IE10+
    window.navigator.msSaveOrOpenBlob(blob, filename);
  } else {
    // Others
    const a = document.createElement('a'),
      url = URL.createObjectURL(blob);
    a.href = url;
    a.download = filename;
    document.body.appendChild(a);
    a.click();
    setTimeout(function () {
      document.body.removeChild(a);
      window.URL.revokeObjectURL(url);
    }, 0);
  }
}

export function readFile(file) {
  return new Promise(function (resolve, reject) {
    const fileReader = new FileReader();
    fileReader.readAsText(file, 'UTF-8');
    fileReader.onload = function () {
      resolve(fileReader.result);
    };
    fileReader.onerror = function (error) {
      reject(error);
    };
  });
}
