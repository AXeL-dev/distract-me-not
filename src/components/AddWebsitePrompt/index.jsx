import React, { Component } from 'react';
import { Pane, toaster } from 'evergreen-ui';
import { translate } from 'helpers/i18n';
import { TextField } from 'components';
import { Mode, blockUrl } from 'helpers/block';
import { isUrl } from 'helpers/url';
import queryString from 'query-string';

export class AddWebsitePrompt extends Component {
  constructor(props) {
    super(props);
    const params = props.location ? queryString.parse(props.location.search) : {};
    this.mode = params.mode || Mode.blacklist;
    this.tabId = +params.tabId || null;
    this.state = {
      url: params.url || '',
      disabled: false,
    };
  }

  componentDidMount() {
    document.addEventListener('keyup', this.handleKeyUp);
  }

  componentWillUnmount() {
    document.removeEventListener('keyup', this.handleKeyUp);
  }

  handleKeyUp = (event) => {
    if (event.code === 'Escape') {
      window.close();
    }
  };

  handleChange = (event) => {
    this.setState({ url: event.target.value });
  };

  handleSubmit = async () => {
    if (!isUrl(this.state.url)) {
      toaster.danger(translate('urlIsNotValid'), {
        id: 'add-website-toaster',
      });
    } else {
      this.setState({ disabled: true });
      try {
        const blocked = await blockUrl(this.state.url, this.mode, this.tabId);
        if (blocked) {
          window.close();
        } else {
          throw new Error(translate('urlAlreadyExists'));
        }
      } catch (error) {
        toaster.danger(error.message, { id: 'add-website-toaster' });
        this.setState({ disabled: false });
      }
    }
  };

  render() {
    return (
      <Pane
        display="flex"
        flex={1}
        width="100%"
        height="100%"
        minWidth={500}
        minHeight={120}
        padding={30}
        alignItems="center"
        justifyContent="center"
      >
        <TextField
          placeholder={translate('urlExample')}
          hint={translate('addWebsiteHint')}
          value={this.state.url}
          hasButton={true}
          buttonLabel={translate('add')}
          onSubmit={this.handleSubmit}
          onChange={this.handleChange}
          disabled={this.state.disabled}
          required
        />
      </Pane>
    );
  }
}
