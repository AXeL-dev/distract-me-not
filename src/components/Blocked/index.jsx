import React, { Component, Fragment } from 'react';
import {
  Pane,
  Dialog,
  RadioGroup,
  Button,
  UnlockIcon,
  toaster,
  DuplicateIcon,
} from 'evergreen-ui';
import { translate } from 'helpers/i18n';
import { storage, sendMessage } from 'helpers/webext';
import { debug, isDevEnv, isProdEnv } from 'helpers/debug';
import { isUrl, getValidUrl } from 'helpers/url';
import {
  UnblockOptions,
  isPageReloaded,
  defaultUnblockSettings,
  defaultBlockSettings,
} from 'helpers/block';
import { NumberInput, PasswordPrompt } from 'components';
import queryString from 'query-string';
import copy from 'copy-to-clipboard';
import './styles.scss';

export class Blocked extends Component {
  constructor(props) {
    super(props);
    this.url = props.location ? queryString.parse(props.location.search).url : null;
    if (this.url && isUrl(this.url)) {
      this.url = decodeURIComponent(this.url);
      this.url = getValidUrl(this.url);
    }
    if (isDevEnv && !this.url) {
      this.url = 'https://www.example.com';
    }
    debug.log('url', this.url);
    const defaultUnblockTime = 10; // min
    this.state = {
      message: props.message || translate('defaultBlockingMessage'),
      isBlank: props.isBlank === undefined ? isProdEnv : props.isBlank, // isBlank should be true by default in prod
      hasUnblockButton: props.hasUnblockButton || isDevEnv,
      displayBlockedLink:
        props.displayBlockedLink || isDevEnv || defaultBlockSettings.displayBlockedLink,
      unblockDialog: {
        isShown: false,
        options: this.getUnblockOptions(defaultUnblockTime),
        selected: UnblockOptions.unblockOnce,
        time: defaultUnblockTime,
        requirePassword: props.requirePassword || false,
      },
    };
  }

  getUnblockOptions = (time) => {
    return [
      {
        label: translate('unblockOnce'),
        value: UnblockOptions.unblockOnce,
      },
      {
        label: (
          <Pane display="flex" alignItems="center" gap={10}>
            <span>{translate('unblockFor')}</span>
            <NumberInput
              min={1}
              max={720}
              width={65}
              value={time}
              onChange={(value) =>
                this.updateUnblockDialogState({
                  selected: UnblockOptions.unblockForWhile,
                  time: value,
                })
              }
            />
            <span>{translate('minutes')}</span>
          </Pane>
        ),
        value: UnblockOptions.unblockForWhile,
      },
    ];
  };

  componentDidMount() {
    if (isPageReloaded()) {
      debug.log('page reloaded!');
      sendMessage('isUrlStillBlocked', this.url).then((isUrlStillBlocked) => {
        // Redirect to blocked url if no longer blocked
        if (isUrlStillBlocked === false && this.url) {
          sendMessage('redirectSenderTab', this.url);
          return;
        }
      });
    }
    storage
      .get({
        message: this.state.message,
        displayBlankPage: this.state.isBlank,
        displayBlockedLink: this.state.displayBlockedLink,
        unblock: {
          isEnabled: isDevEnv || defaultUnblockSettings.isEnabled,
          requirePassword: defaultUnblockSettings.requirePassword,
        },
        password: {
          isEnabled: false,
        },
      })
      .then((items) => {
        if (items) {
          this.setState({
            message: items.message.length ? items.message : this.state.message,
            isBlank: items.displayBlankPage,
            displayBlockedLink: items.displayBlockedLink,
            hasUnblockButton: items.unblock.isEnabled,
            unblockDialog: {
              ...this.state.unblockDialog,
              requirePassword:
                items.unblock.isEnabled &&
                items.unblock.requirePassword &&
                items.password.isEnabled,
            },
          });
        }
      });
  }

  updateUnblockDialogState = (state) => {
    this.setState({
      unblockDialog: {
        ...this.state.unblockDialog,
        ...state,
      },
    });
  };

  closeUnblockDialog = () => {
    this.updateUnblockDialogState({ isShown: false });
  };

  openUnblockDialog = () => {
    this.updateUnblockDialogState({ isShown: true });
  };

  unblock = () => {
    this.closeUnblockDialog();
    const params = {
      url: this.url,
      option: this.state.unblockDialog.selected,
      time: this.state.unblockDialog.time,
    };
    debug.log('unblocking:', params);
    if (this.url) {
      //window.location.replace(this.url);
      sendMessage('unblockSenderTab', params);
    }
  };

  copyBlockedLink = () => {
    if (copy(this.url)) {
      toaster.success(translate('copiedToClipboard'), {
        id: 'blocked-toaster',
      });
    }
  };

  render() {
    return (
      <Fragment>
        {!this.state.isBlank && (
          <Fragment>
            <div className="distract-cursor distract-select distract-overlay-container">
              <div className="distract-cursor distract-select distract-overlay">
                <div className="distract-cursor distract-select distract-info-container">
                  <span className="distract-cursor distract-select distract-overlay-top-text">
                    {this.state.message}
                  </span>
                  {this.state.displayBlockedLink && (
                    <span className="distract-blocked-link">
                      <input type="text" value={this.url || ''} readOnly />
                      <button
                        className="copy"
                        title={translate('copy')}
                        onClick={this.copyBlockedLink}
                      >
                        <DuplicateIcon />
                      </button>
                    </span>
                  )}
                  <div className="distract-cursor distract-select distract-overlay-img"></div>
                  {this.state.hasUnblockButton && (
                    <button className="unblock" onClick={this.openUnblockDialog}>
                      {translate('unblock')}
                    </button>
                  )}
                </div>
              </div>
            </div>
            <Dialog
              isShown={this.state.unblockDialog.isShown}
              onCloseComplete={this.closeUnblockDialog}
              shouldCloseOnOverlayClick={false}
              //hasHeader={false}
              hasFooter={false}
              topOffset="40vmin"
              width={400}
              containerProps={{
                className: 'unblock-dialog',
              }}
            >
              <Pane width="95%" margin="auto">
                <RadioGroup
                  size={16}
                  value={this.state.unblockDialog.selected}
                  options={this.state.unblockDialog.options}
                  onChange={(event) =>
                    this.updateUnblockDialogState({
                      selected: event.target.value,
                    })
                  }
                />
                {this.state.unblockDialog.requirePassword ? (
                  <PasswordPrompt
                    hasHeader={false}
                    hasFooter={false}
                    minWidth="auto"
                    minHeight={50}
                    inputWidth="100%"
                    maxInputWidth="auto"
                    onSuccess={this.unblock}
                  />
                ) : (
                  <Pane
                    display="flex"
                    alignItems="center"
                    justifyContent="center"
                    marginTop={20}
                  >
                    <Button
                      height={32}
                      appearance="primary"
                      iconBefore={UnlockIcon}
                      onClick={this.unblock}
                    >
                      {translate('unblock')}
                    </Button>
                  </Pane>
                )}
              </Pane>
            </Dialog>
          </Fragment>
        )}
      </Fragment>
    );
  }
}
