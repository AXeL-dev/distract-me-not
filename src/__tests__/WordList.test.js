import React from 'react';
import { render, screen, fireEvent, within, waitFor } from "@testing-library/react";
import * as fileHelper from "helpers/file";
import { WordList } from "components";
import copy from 'copy-to-clipboard';

const keywords = [
  'foo',
  'bar',
];

it('renders correctly', () => {
  const { container, asFragment } = render(<WordList list={keywords} />);
  for (let keyword of keywords) {
    expect(container).toHaveTextContent(keyword);
  }
  expect(asFragment()).toMatchSnapshot();
});

it('adds a keyword to list', () => {
  const keyword = 'test';
  const { container } = render(<WordList list={[]} />);
  const list = container.querySelector('div[data-evergreen-table-body="true"]');
  // add keyword
  const addInput = screen.getByPlaceholderText('keywordExample', { selector: 'input[type="text"]' });
  fireEvent.change(addInput, { target: { value: keyword } });
  fireEvent.keyDown(addInput, { key: 'Enter' });
  // verify
  expect(list).toHaveTextContent(keyword);
});

it('edits a keyword', async () => {
  const keyword = {
    value: 'foo',
    replacement: 'bar'
  };
  const { container } = render(<WordList list={[keyword.value]} />);
  const list = container.querySelector('div[data-evergreen-table-body="true"]');
  // click on more button
  const moreButton = screen.getByTestId('more-button');
  fireEvent.click(moreButton);
  // click on edit button
  const editButton = await waitFor(() => screen.getByRole('menuitem', { name: /edit/i }));
  fireEvent.click(editButton);
  // change keyword value
  const editDialog = await waitFor(() => screen.getByRole('dialog', { selector: '[data-state="entered"]' }));
  const editInput = within(editDialog).getByPlaceholderText('keywordExample');
  fireEvent.change(editInput, { target: { value: keyword.replacement } });
  // save
  const saveButton = screen.getByRole('button', { name: /edit/i });
  fireEvent.click(saveButton);
  // verify
  expect(list).toHaveTextContent(keyword.replacement);
});

it('deletes a keyword', async () => {
  const keyword = 'test';
  const { container } = render(<WordList list={[keyword]} />);
  const list = container.querySelector('div[data-evergreen-table-body="true"]');
  // check if keyword is there
  expect(list).toHaveTextContent(keyword);
  // click on more button
  const moreButton = screen.getByTestId('more-button');
  fireEvent.click(moreButton);
  // click on delete button
  const deleteButton = await waitFor(() => screen.getByRole('menuitem', { name: /delete/i }));
  fireEvent.click(deleteButton);
  // verify
  expect(list).not.toHaveTextContent(keyword);
});

it('copies a keyword to clipboard', async () => {
  render(<WordList list={[keywords[0]]} />);
  // click on more button
  const moreButton = screen.getByTestId('more-button');
  fireEvent.click(moreButton);
  // click on copy button
  const copyButton = await waitFor(() => screen.getByRole('menuitem', { name: /copy/i }));
  fireEvent.click(copyButton);
  // verify
  expect(copy).toHaveBeenCalledTimes(1);
});

it('filters keywords', () => {
  const filter = keywords[0];
  const { container } = render(<WordList list={keywords} />);
  const list = container.querySelector('div[data-evergreen-table-body="true"]');
  // filter by keyword
  const filterInput = screen.getByPlaceholderText('filter...', { selector: 'input[type="text"]' });
  fireEvent.change(filterInput, { target: { value: filter } });
  // verify
  for (let keyword of keywords) {
    if (keyword === filter) continue;
    expect(list).not.toHaveTextContent(keyword);
  }
  expect(list).toHaveTextContent(filter);
});

it('sorts keywords in ascending order', async () => {
  const { container } = render(<WordList list={keywords} />);
  const list = container.querySelector('div[data-evergreen-table-body="true"]');
  // click on sort button
  const sortButton = screen.getByTestId('sort-button');
  fireEvent.click(sortButton);
  // choose asc order
  const ascButton = await waitFor(() => screen.getByRole('menuitemradio', { name: /ascending/i }));
  fireEvent.click(ascButton);
  // verify
  const listOrder = within(list).getAllByTestId('keyword').map(keyword => keyword.innerHTML);
  const expectedOrder = keywords.slice().sort((a, b) => a.localeCompare(b)); // asc
  expect(listOrder).toEqual(expectedOrder);
});

it('sorts keywords in descending order', async () => {
  const { container } = render(<WordList list={keywords} />);
  const list = container.querySelector('div[data-evergreen-table-body="true"]');
  // click on sort button
  const sortButton = screen.getByTestId('sort-button');
  fireEvent.click(sortButton);
  // choose desc order
  const descButton = await waitFor(() => screen.getByRole('menuitemradio', { name: /descending/i }));
  fireEvent.click(descButton);
  // verify
  const listOrder = within(list).getAllByTestId('keyword').map(keyword => keyword.innerHTML);
  const expectedOrder = keywords.slice().sort((a, b) => b.localeCompare(a)); // desc
  expect(listOrder).toEqual(expectedOrder);
});

it('imports keywords', async () => {
  const { container } = render(<WordList list={[]} />);
  const list = container.querySelector('div[data-evergreen-table-body="true"]');
  // trigger file input change event
  const fileInput = screen.getByTestId('file-input');
  const file = new File(keywords, 'blacklist.txt', { type: 'text/plain' });
  fireEvent.change(fileInput, { target: { files: [file] } });
  // verify
  for (let keyword of keywords) {
    await waitFor(() => expect(list).toHaveTextContent(keyword));
  }
});

it('exports keywords', async () => {
  const exportFilename = 'keywords.txt';
  render(<WordList list={keywords} exportFilename={exportFilename} />);
  // spy on file helper download function
  const downloadFn = jest.spyOn(fileHelper, 'download');
  global.URL.createObjectURL = jest.fn(); // fix error: URL.createObjectURL is not a function
  global.Blob = function(content, options) { // allow us to compare Blobs
    return {
      content: content[0].split("\n"),
      options
    };
  };
  // click on more button
  const moreButton = screen.getByTestId('list-more-button');
  fireEvent.click(moreButton);
  // click on export button
  const exportButton = await waitFor(() => screen.getByRole('menuitem', { name: /export/i }));
  fireEvent.click(exportButton);
  // verify
  expect(downloadFn).toHaveBeenCalledWith(
    {
      content: keywords,
      options: {
        type: 'text/plain'
      }
    }, 
    exportFilename
  );
});
