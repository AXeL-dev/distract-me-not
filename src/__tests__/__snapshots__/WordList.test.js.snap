// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<DocumentFragment>
  <div
    class="ub-b-top_1px-solid-E6E8F0 ub-b-rgt_1px-solid-E6E8F0 ub-b-btm_1px-solid-E6E8F0 ub-b-lft_1px-solid-E6E8F0 ub-bblr_4px ub-bbrr_4px ub-btlr_4px ub-btrr_4px ub-box-szg_border-box"
  >
    <div
      class="css-ytz3eb ub-bg_wivy90 ub-b-btm_1px-solid-E6E8F0 ub-pr_0px ub-h_32px ub-dspl_flex ub-flx-srnk_0 ub-fnt-sze_12px ub-f-wght_600 ub-ln-ht_16px ub-ltr-spc_0 ub-fnt-fam_b77syt ub-color_696f8c ub-txt-trns_uppercase ub-pb_0px ub-pl_0px ub-pt_0px ub-box-szg_border-box"
    >
      <div
        class="css-ber32g ub-box-szg_border-box ub-flx_1 ub-dspl_flex ub-algn-itms_center ub-flx-srnk_0 ub-ovflw-x_visible ub-ovflw-y_visible ub-pl_12px ub-pr_12px"
      >
        <div
          class="ub-dspl_inline-flex ub-ml_2px ub-mr_10px ub-box-szg_border-box"
        >
          <svg
            class="css-1de2wot ub-w_12px ub-h_12px ub-box-szg_border-box"
            data-icon="search"
            viewBox="0 0 16 16"
          >
            <path
              d="M15.55 13.43l-2.67-2.68a6.94 6.94 0 001.11-3.76c0-3.87-3.13-7-7-7s-7 3.13-7 7 3.13 7 7 7c1.39 0 2.68-.42 3.76-1.11l2.68 2.67a1.498 1.498 0 102.12-2.12zm-8.56-1.44c-2.76 0-5-2.24-5-5s2.24-5 5-5 5 2.24 5 5-2.24 5-5 5z"
              fill-rule="evenodd"
            />
          </svg>
        </div>
        <input
          class="css-q4k7k7 ub-fnt-sze_12px ub-f-wght_500 ub-ln-ht_16px ub-ltr-spc_0 ub-fnt-fam_b77syt ub-color_474d66 ub-flx_1 ub-ml_-2px ub-pl_0px ub-box-szg_border-box"
          placeholder="filter..."
          spellcheck="true"
          value=""
        />
      </div>
      <div
        class="css-ber32g ub-box-szg_border-box ub-flx_none ub-dspl_flex ub-algn-itms_center ub-flx-srnk_0 ub-ovflw-x_visible ub-ovflw-y_visible ub-pl_12px ub-pr_12px"
      >
        <button
          aria-haspopup="true"
          class="css-m3k0tf ub-pst_relative ub-f-wght_500 ub-dspl_inline-flex ub-algn-itms_center ub-flx-wrap_nowrap ub-just-cnt_center ub-txt-deco_none ub-ver-algn_middle ub-b-btm_none ub-b-lft_none ub-b-rgt_none ub-b-top_none ub-otln_iu2jf4 ub-usr-slct_none ub-crsr_pointer ub-wht-spc_nowrap ub-fnt-fam_b77syt ub-bg-clr_transparent ub-bblr_4px ub-bbrr_4px ub-btlr_4px ub-btrr_4px ub-pl_4px ub-pr_4px ub-ml_-4px ub-mr_-4px ub-pb_2px ub-pt_2px ub-mb_-2px ub-mt_-2px ub-color_474d66 ub-fnt-sze_12px ub-ln-ht_16px ub-box-szg_border-box"
          data-testid="sort-button"
          role="button"
          type="button"
        >
          <div
            class="ub-dspl_inline-flex ub-ml_2px ub-box-szg_border-box"
          >
            <svg
              class="css-178dftq ub-w_12px ub-h_12px ub-box-szg_border-box"
              data-icon="caret-down"
              viewBox="0 0 16 16"
            >
              <path
                d="M12 6.5c0-.28-.22-.5-.5-.5h-7a.495.495 0 00-.37.83l3.5 4c.09.1.22.17.37.17s.28-.07.37-.17l3.5-4c.08-.09.13-.2.13-.33z"
                fill-rule="evenodd"
              />
            </svg>
          </div>
        </button>
      </div>
      <div
        class="css-ber32g ub-box-szg_border-box ub-flx_none ub-dspl_flex ub-algn-itms_center ub-flx-srnk_0 ub-ovflw-x_visible ub-ovflw-y_visible ub-pl_12px ub-pr_12px ub-w_48px"
      >
        <button
          aria-haspopup="true"
          class="css-1tyjqqv ub-pst_relative ub-f-wght_400 ub-dspl_inline-flex ub-algn-itms_center ub-flx-wrap_nowrap ub-just-cnt_center ub-txt-deco_none ub-ver-algn_middle ub-b-btm_1px-solid-transparent ub-b-lft_1px-solid-transparent ub-b-rgt_1px-solid-transparent ub-b-top_1px-solid-transparent ub-otln_iu2jf4 ub-usr-slct_none ub-crsr_pointer ub-wht-spc_nowrap ub-fnt-fam_b77syt ub-bblr_4px ub-bbrr_4px ub-btlr_4px ub-btrr_4px ub-color_474d66 ub-tstn_n1akt6 ub-h_24px ub-min-w_24px ub-fnt-sze_12px ub-ln-ht_16px ub-pl_0px ub-pr_0px ub-bg-clr_transparent ub-flx_none ub-w_24px ub-ltr-spc_0 ub-box-szg_border-box"
          data-testid="list-more-button"
          role="button"
        >
          <div
            class="ub-dspl_inline-flex ub-box-szg_border-box"
          >
            <svg
              class="css-178dftq ub-w_12px ub-h_12px ub-box-szg_border-box"
              data-icon="more"
              viewBox="0 0 16 16"
            >
              <path
                d="M2 6.03a2 2 0 100 4 2 2 0 100-4zM14 6.03a2 2 0 100 4 2 2 0 100-4zM8 6.03a2 2 0 100 4 2 2 0 100-4z"
                fill-rule="evenodd"
              />
            </svg>
          </div>
        </button>
      </div>
       
      <div
        aria-hidden="true"
        style="position: fixed; top: -500px; left: -500px; width: 100px; overflow-y: scroll;"
      >
        <div />
      </div>
    </div>
    <div
      class="ub-h_240px ub-flx_1 ub-ovflw-x_hidden ub-ovflw-y_hidden ub-box-szg_border-box"
      data-evergreen-table-body="true"
    >
      <div
        style="overflow: auto; will-change: transform; height: 240px; width: 100%;"
      >
        <div
          style="position: relative; width: 100%; min-height: 100%; height: 144px;"
        >
          <div
            class="css-15su02r ub-b-btm_1px-solid-edeff5 ub-h_64px ub-dspl_flex ub-otln_iu2jf4 ub-txt-deco_none ub-bg-clr_white ub-box-szg_border-box"
            style="position: absolute; top: 0px; left: 0px; width: 100%; height: 48px;"
          >
            <div
              class="css-ber32g ub-box-szg_border-box ub-flx_1 ub-dspl_flex ub-algn-itms_center ub-flx-srnk_0 ub-ovflw-x_hidden ub-ovflw-y_hidden ub-pl_12px ub-pr_12px"
            >
              <span
                class="ub-fnt-sze_12px ub-f-wght_500 ub-ln-ht_16px ub-ltr-spc_0 ub-fnt-fam_b77syt ub-color_474d66 ub-ml_8px ub-box-szg_border-box"
                data-testid="keyword"
              >
                foo
              </span>
            </div>
            <div
              class="css-ber32g ub-box-szg_border-box ub-flx_none ub-dspl_flex ub-algn-itms_center ub-flx-srnk_0 ub-ovflw-x_hidden ub-ovflw-y_hidden ub-pl_12px ub-pr_12px ub-w_48px"
            >
              <button
                aria-haspopup="true"
                class="css-1tyjqqv ub-pst_relative ub-f-wght_400 ub-dspl_inline-flex ub-algn-itms_center ub-flx-wrap_nowrap ub-just-cnt_center ub-txt-deco_none ub-ver-algn_middle ub-b-btm_1px-solid-transparent ub-b-lft_1px-solid-transparent ub-b-rgt_1px-solid-transparent ub-b-top_1px-solid-transparent ub-otln_iu2jf4 ub-usr-slct_none ub-crsr_pointer ub-wht-spc_nowrap ub-fnt-fam_b77syt ub-bblr_4px ub-bbrr_4px ub-btlr_4px ub-btrr_4px ub-color_474d66 ub-tstn_n1akt6 ub-h_24px ub-min-w_24px ub-fnt-sze_12px ub-ln-ht_16px ub-pl_0px ub-pr_0px ub-bg-clr_transparent ub-flx_none ub-w_24px ub-ltr-spc_0 ub-box-szg_border-box"
                data-testid="more-button"
                role="button"
              >
                <div
                  class="ub-dspl_inline-flex ub-box-szg_border-box"
                >
                  <svg
                    class="css-178dftq ub-w_12px ub-h_12px ub-box-szg_border-box"
                    data-icon="more"
                    viewBox="0 0 16 16"
                  >
                    <path
                      d="M2 6.03a2 2 0 100 4 2 2 0 100-4zM14 6.03a2 2 0 100 4 2 2 0 100-4zM8 6.03a2 2 0 100 4 2 2 0 100-4z"
                      fill-rule="evenodd"
                    />
                  </svg>
                </div>
              </button>
            </div>
          </div>
          <div
            class="css-15su02r ub-b-btm_1px-solid-edeff5 ub-h_64px ub-dspl_flex ub-otln_iu2jf4 ub-txt-deco_none ub-bg-clr_white ub-box-szg_border-box"
            style="position: absolute; top: 48px; left: 0px; width: 100%; height: 48px;"
          >
            <div
              class="css-ber32g ub-box-szg_border-box ub-flx_1 ub-dspl_flex ub-algn-itms_center ub-flx-srnk_0 ub-ovflw-x_hidden ub-ovflw-y_hidden ub-pl_12px ub-pr_12px"
            >
              <span
                class="ub-fnt-sze_12px ub-f-wght_500 ub-ln-ht_16px ub-ltr-spc_0 ub-fnt-fam_b77syt ub-color_474d66 ub-ml_8px ub-box-szg_border-box"
                data-testid="keyword"
              >
                bar
              </span>
            </div>
            <div
              class="css-ber32g ub-box-szg_border-box ub-flx_none ub-dspl_flex ub-algn-itms_center ub-flx-srnk_0 ub-ovflw-x_hidden ub-ovflw-y_hidden ub-pl_12px ub-pr_12px ub-w_48px"
            >
              <button
                aria-haspopup="true"
                class="css-1tyjqqv ub-pst_relative ub-f-wght_400 ub-dspl_inline-flex ub-algn-itms_center ub-flx-wrap_nowrap ub-just-cnt_center ub-txt-deco_none ub-ver-algn_middle ub-b-btm_1px-solid-transparent ub-b-lft_1px-solid-transparent ub-b-rgt_1px-solid-transparent ub-b-top_1px-solid-transparent ub-otln_iu2jf4 ub-usr-slct_none ub-crsr_pointer ub-wht-spc_nowrap ub-fnt-fam_b77syt ub-bblr_4px ub-bbrr_4px ub-btlr_4px ub-btrr_4px ub-color_474d66 ub-tstn_n1akt6 ub-h_24px ub-min-w_24px ub-fnt-sze_12px ub-ln-ht_16px ub-pl_0px ub-pr_0px ub-bg-clr_transparent ub-flx_none ub-w_24px ub-ltr-spc_0 ub-box-szg_border-box"
                data-testid="more-button"
                role="button"
              >
                <div
                  class="ub-dspl_inline-flex ub-box-szg_border-box"
                >
                  <svg
                    class="css-178dftq ub-w_12px ub-h_12px ub-box-szg_border-box"
                    data-icon="more"
                    viewBox="0 0 16 16"
                  >
                    <path
                      d="M2 6.03a2 2 0 100 4 2 2 0 100-4zM14 6.03a2 2 0 100 4 2 2 0 100-4zM8 6.03a2 2 0 100 4 2 2 0 100-4z"
                      fill-rule="evenodd"
                    />
                  </svg>
                </div>
              </button>
            </div>
          </div>
          <div
            class="css-15su02r ub-b-btm_1px-solid-edeff5 ub-h_64px ub-dspl_flex ub-otln_iu2jf4 ub-txt-deco_none ub-bg-clr_white ub-box-szg_border-box"
            style="position: absolute; top: 96px; left: 0px; width: 100%; height: 48px;"
          >
            <div
              class="css-ber32g ub-box-szg_border-box ub-flx_1 ub-dspl_flex ub-algn-itms_center ub-flx-srnk_0 ub-ovflw-x_hidden ub-ovflw-y_hidden ub-pl_12px ub-pr_12px"
            >
              <span
                class="ub-fnt-sze_12px ub-f-wght_500 ub-ln-ht_16px ub-ltr-spc_0 ub-fnt-fam_b77syt ub-color_474d66 ub-ml_8px ub-box-szg_border-box"
                data-testid="keyword"
              >
                /doo.*/i
              </span>
            </div>
            <div
              class="css-ber32g ub-box-szg_border-box ub-flx_none ub-dspl_flex ub-algn-itms_center ub-flx-srnk_0 ub-ovflw-x_hidden ub-ovflw-y_hidden ub-pl_12px ub-pr_12px ub-w_48px"
            >
              <button
                aria-haspopup="true"
                class="css-1tyjqqv ub-pst_relative ub-f-wght_400 ub-dspl_inline-flex ub-algn-itms_center ub-flx-wrap_nowrap ub-just-cnt_center ub-txt-deco_none ub-ver-algn_middle ub-b-btm_1px-solid-transparent ub-b-lft_1px-solid-transparent ub-b-rgt_1px-solid-transparent ub-b-top_1px-solid-transparent ub-otln_iu2jf4 ub-usr-slct_none ub-crsr_pointer ub-wht-spc_nowrap ub-fnt-fam_b77syt ub-bblr_4px ub-bbrr_4px ub-btlr_4px ub-btrr_4px ub-color_474d66 ub-tstn_n1akt6 ub-h_24px ub-min-w_24px ub-fnt-sze_12px ub-ln-ht_16px ub-pl_0px ub-pr_0px ub-bg-clr_transparent ub-flx_none ub-w_24px ub-ltr-spc_0 ub-box-szg_border-box"
                data-testid="more-button"
                role="button"
              >
                <div
                  class="ub-dspl_inline-flex ub-box-szg_border-box"
                >
                  <svg
                    class="css-178dftq ub-w_12px ub-h_12px ub-box-szg_border-box"
                    data-icon="more"
                    viewBox="0 0 16 16"
                  >
                    <path
                      d="M2 6.03a2 2 0 100 4 2 2 0 100-4zM14 6.03a2 2 0 100 4 2 2 0 100-4zM8 6.03a2 2 0 100 4 2 2 0 100-4z"
                      fill-rule="evenodd"
                    />
                  </svg>
                </div>
              </button>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    class="ub-mt_16px ub-box-szg_border-box"
  >
    <div
      class="ub-dspl_flex ub-box-szg_border-box"
    >
      <div
        class="ub-dspl_flex ub-algn-itms_center ub-flx_1 ub-box-szg_border-box"
      >
        <input
          aria-invalid="false"
          class="css-12rdf6g ub-w_100prcnt ub-fnt-fam_b77syt ub-b-btm_1px-solid-transparent ub-b-lft_1px-solid-transparent ub-b-rgt_1px-solid-transparent ub-b-top_1px-solid-transparent ub-otln_iu2jf4 ub-txt-deco_none ub-bblr_4px ub-bbrr_0px ub-btlr_4px ub-btrr_0px ub-ln-ht_16px ub-fnt-sze_12px ub-color_474d66 ub-pl_12px ub-pr_12px ub-tstn_n1akt6 ub-h_32px ub-bg-clr_white ub-b-btm-clr_d8dae5 ub-b-lft-clr_d8dae5 ub-b-rgt-clr_d8dae5 ub-b-top-clr_d8dae5 ub-box-szg_border-box"
          placeholder="keywordExample"
          required=""
          spellcheck="true"
          type="text"
          value=""
        />
      </div>
      <div
        class="ub-dspl_flex ub-algn-itms_center ub-ml_-1px ub-box-szg_border-box"
      >
        <button
          class="css-elp417 ub-pst_relative ub-f-wght_500 ub-dspl_inline-flex ub-algn-itms_center ub-flx-wrap_nowrap ub-just-cnt_center ub-txt-deco_none ub-ver-algn_middle ub-b-btm_1px-solid-transparent ub-b-lft_1px-solid-transparent ub-b-rgt_1px-solid-transparent ub-b-top_1px-solid-transparent ub-otln_iu2jf4 ub-usr-slct_none ub-crsr_pointer ub-wht-spc_nowrap ub-fnt-fam_b77syt ub-bblr_0px ub-bbrr_4px ub-btlr_0px ub-btrr_4px ub-color_white ub-tstn_n1akt6 ub-h_32px ub-min-w_32px ub-fnt-sze_12px ub-ln-ht_32px ub-pl_16px ub-pr_16px ub-bg-clr_3366FF ub-b-btm-clr_3366FF ub-b-lft-clr_3366FF ub-b-rgt-clr_3366FF ub-b-top-clr_3366FF ub-box-szg_border-box"
        >
          add
        </button>
      </div>
    </div>
    <p
      class="ub-fnt-fam_b77syt ub-color_696f8c ub-mt_6px ub-mb_0px ub-fnt-sze_12px ub-f-wght_400 ub-ln-ht_18px ub-ltr-spc_0 ub-box-szg_border-box"
    >
      <span
        class=""
      >
        addKeywordHint
      </span>
    </p>
  </div>
  <input
    accept=".txt"
    class="hidden"
    data-testid="file-input"
    type="file"
  />
</DocumentFragment>
`;
